# 01. Importing libraries

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [2]:
#Create path
path = r'C:\Users\hnguy\OneDrive\Documents\Instacart Basket Analysis'

In [4]:
#Import orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

In [5]:
#Import products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 03. Data Consistency

In [6]:
#Creating dataframe for mixed types
df_test = pd.DataFrame()

In [7]:
#Creating a mixed type data colum for the dataframe
df_test['mix'] = ['a', 'b', 1, True]

In [8]:
#Check for any mixed type data
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)

mix


In [9]:
#Change the mixed type to all string
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
df_test

mix
0     a
1     b
2     1
3  True

# Missing Values

In [11]:
#Check for missing values in df_prods data
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
#Create a dataframe for all the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
df_prods.shape

(49693, 5)

In [16]:
#Create a clean dataframe that does not consist of the missing data
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
df_prods_clean.shape

(49677, 5)

# Duplicate Values

In [18]:
#Create a dataframe that shows all the duplicate values
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [20]:
#Create a clean dataframe that does not contain the duplicate values
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task 4.5

Run the df.describe() function on your df_prods dataframe. Check whether anything about the data looks off or should be investigated further.

In [22]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The mean is about 10 but the maximum value is 99999. This is a large outlier and could be an error in the data.

Check for mixed-type data in your df_ords dataframe.

In [23]:
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

No mixed-type data found.

Run a check for missing values in your df_ords dataframe.

In [24]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_days_of_week            0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values for the "days_since_prior_order" column. This could have happened because some customers are new customers and have not made any orders before.

We can create a new column for new customers to help distinguish them from the old customers.

In [25]:
df_ords_clean = df_ords

In [27]:
df_ords_clean['new_customer'] = df_ords['days_since_prior_order'].isnull() == False

In [28]:
df_ords_clean

Unnamed: 0  order_id  user_id  order_number  orders_days_of_week  \
0                 0   2539329        1             1                    2   
1                 1   2398795        1             2                    3   
2                 2    473747        1             3                    3   
3                 3   2254736        1             4                    4   
4                 4    431534        1             5                    4   
...             ...       ...      ...           ...                  ...   
3421078     3421078   2266710   206209            10                    5   
3421079     3421079   1854736   206209            11                    4   
3421080     3421080    626363   206209            12                    1   
3421081     3421081   2977660   206209            13                    1   
3421082     3421082    272231   206209            14                    6   

         order_hour_of_day  days_since_prior_order  new_customer  
0                        8                     NaN         False  
1                        7                    15.0          True  
2                       12                    21.0          True  
3                        7                    29.0          True  
4                       15                    28.0          True  
...                    ...                     ...           ...  
3421078                 18                    29.0          True  
3421079                 10                    30.0          True  
3421080                 12                    18.0          True  
3421081                 12                     7.0          True  
3421082                 14                    30.0          True  

[3421083 rows x 8 columns]

Run a check for duplicate values in your df_ords data.

In [29]:
df_dupes = df_ords_clean[df_ords_clean.duplicated()]

In [30]:
df_dupes

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_days_of_week, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

There are no duplicates in the df_ords data.

Export your final, clean df_prods and df_ords data as ".csv" files in your "Prepared Data" folder

In [31]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_clean.csv'))

In [ ]:
df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders.clean.'))